vowpalwabbit/core/src/reductions/plt.cc

```diff
@@ -323,12 +341,11 @@ void save_load_tree(plt& p, io_buf& model_file, bool read, bool text)
 {
   if (model_file.num_files() > 0)
   {
-    bool resume = p.all->save_resume;
     std::stringstream msg;
-    msg << ":" << resume << "\n";
+    bool resume = p.all->save_resume;
     bin_text_read_write_fixed(model_file, reinterpret_cast<char*>(&resume), sizeof(resume), read, msg, text);
 
-    if (resume && !p.all->weights.adaptive)
```

Comment:
What's the reasoning behind this change?

Reply:
It seems to me that the previous version is wrong because during the prediction `p.all->weights.adaptive` is always false (at least if no additional arguments are provided), so if the model was trained with `p.all->weights.adaptive == true` it's later loaded in the wrong way. Removing this and only depending on the resume variable fix the problem + in the current version this `node_time` array that is saved is used for all optimizer types (see `learn_node` function).

Reply:
I think this is better now since the current version does not depend on optimizer-specifics, even if the optimizer does not use sd->t (like adaptive), writing/reading one additional float per tree node is a small price for more general code.

Reply:
And since `p.all->weights.adaptive == true` during prediction, removing it still keeps compatibility with older models.

Reply:
I just spent some time to fully understand this situation. This previous `!p.all->weights.adaptive == true` check is actually quite troublesome. If loading an older model with this `save_load` function then most code paths are invalid. The default (no optimizer flags) will also be invalid. I think the best course of action is to disallow loading of old models with an error message explaining why, and providing an override flag if a user must upgrade a model and understands the risk which is only present in VW 9.7 and removed after. 

| Model produced by VW | Adaptive value when saving | Adaptive value when loading | Result |
|---|---|---|---|
| <9.7 | false | true | valid |
| <9.7 | true | true | invalid (current default) |
| <9.7 | false | false | valid |
| <9.7 | true | false | invalid |
| >=9.7 | n/a | n/a | valid |

We cannot distinguish between the adaptive value when saving or loading easily in the current system (technically possible but tricky). Additionally, the method by which adaptive becomes true or false is a bit complex.

Additionally, any non-gd base optimizer is actually also undefined behavior because the `p.all->weights.adaptive` value is actually uninitialized if gd is not the base.

Based on all these facts I think the *safest* thing to do is disallow loading of <9.7 plt models but allow an escape hatch *just* for version 9.7 to force the load.

We will need to bump the version to 9.7 in this PR for the check to work.

If all sounds good to you, I am happy to push a commit making these changes.

Reply:
@jackgerrits Thank you for putting your time into this. Yes, your table well summarizes the situation. I agree that disallowing or at least not setting some warning is a good thing in this situation. So I'm happy with your proposal.

Reply:
Great, thanks for confirming! I'll fix this up tomorrow then we should be good to merge

Reply:
Just pushed the change:
- By default we fail for <9.7 models being loaded
- A user can override this with `--plt_force_load_legacy_model` which will be removed in the next release
- I remove the save/load of `resume` as you had originally done in this PR
- `nodes_time` is now unconditionally saved in the model. In addition its size is saved in the model and we do an assert to ensure `t` and `nodes_time.size()` stay in sync.
  - This is a format change due to the removal of `resume` and prefixed size field when writing `nodes_time` and therefore this model *cannot* be loaded in `vw` binaries prior to 9.7